### DISCUSSION TITLE
What do the various metrics for MAP metric mean?

### DISCUSSION THREAD

**Discussion by dhruvbird** [Lightning-AI/torchmetrics#1700] [CompVision] [answered]
On https://torchmetrics.readthedocs.io/en/stable/detection/mean_average_precision.html there are a bunch of metrics reported such as `map_small`, `map_medium`, `map_large`. Is there any place where the interpretation of these reported metrics is explained?

Also, is there any additional documentation around the meaning, interpretation, and use of the following options:
1. `max_detection_thresholds`
2. `iou_thresholds`
3. `rec_thresholds`

Thanks!

**Comment by dhruvbird** [comment]
Okay so it seems like `iou_thresholds` is a list of IoU values to use when computing mAP. i.e. the AP at each of the values in the list is computed, and then averaged (mean is computed) to produce the `mAP` metric. I found this information on [this page](https://pyimagesearch.com/2022/05/02/mean-average-precision-map-using-the-coco-evaluator/)

By default, it seems like the following IoU values are used as thresholds `[0.5, 0.55, 0.60, 0.65, 0.70, 0.75, 0.80, 0.85, 0.90, 0.95]` which is `10` values.

So, if the object detection model is able to predict the bounding boxes but with a lower IoU (say 0.6), it would penalize the `mAP` metric, since it's averaged across a range of values. The `mAP` metric would be high if most of the predicted bounding boxes have a high IoU with the ground truth bounding boxes.

If you want to compute the `AP` metric for all predicted bounding boxes that have an IoU of `0.6` or more, then just pass `[0.6]` as the argument for `iou_thresholds`.

**Comment by Borda** [comment, marked as answer]
@dhruvbird covered `iou_thresholds` well. Let me fill in the remaining fields — these follow the COCO evaluation protocol, and as of v1.9.0 the docs explicitly list them:

| Output Key | What it means |
|---|---|
| `map` | Mean AP averaged across **all** IoU thresholds (0.50:0.05:0.95) and all area sizes |
| `map_50` | AP at IoU threshold = 0.50 only |
| `map_75` | AP at IoU threshold = 0.75 only (stricter localization) |
| `map_small` | AP for **small** objects: area < 32² pixels |
| `map_medium` | AP for **medium** objects: 32² ≤ area ≤ 96² pixels |
| `map_large` | AP for **large** objects: area > 96² pixels |
| `mar_1` | Mean Recall with at most **1** detection per image |
| `mar_10` | Mean Recall with at most **10** detections per image |
| `mar_100` | Mean Recall with at most **100** detections per image |

**The parameters that control these:**

- `iou_thresholds` (default `[0.5, 0.55, ..., 0.95]`): the IoU thresholds for map/mar. `map_50` and `map_75` are always extracted from this range.
- `max_detection_thresholds` (default `[1, 10, 100]`): controls the 1/10/100 in `mar_1`, `mar_10`, `mar_100`. If you change this to `[1, 5, 50]`, you'll get `mar_1`, `mar_5`, `mar_50`.
- `rec_thresholds` (default `[0.0, 0.01, ..., 1.0]`): the recall points at which precision is interpolated to compute AP. More points = smoother curve. Rarely needs changing.
- Area ranges are **hardcoded** to COCO standard (small < 32², medium 32²–96², large > 96²). If all your objects fall in one size bucket, the other `map_*` values will be `-1`.

**Example:**
```python
from torchmetrics.detection import MeanAveragePrecision

metric = MeanAveragePrecision(
    iou_thresholds=[0.5, 0.75],           # only compute at these two IoU
    max_detection_thresholds=[1, 10, 100], # default
    class_metrics=True,                    # also get per-class breakdown
)
```

Docs: [MeanAveragePrecision](https://lightning.ai/docs/torchmetrics/stable/detection/mean_average_precision.html)
